### PR TITLE
feat: Validate workspace specification before accessing parts of it

### LINKED ISSUE #1952: Model specification is accessed before being validated
### Summary

The model specification is currently accessed before validation is done:
https://github.com/scikit-hep/pyhf/blob/a6186405ae5acee472133874e9a477dc41def7df/src/pyhf/mixins.py#L26-L37 

called via https://github.com/scikit-hep/pyhf/blob/a6186405ae5acee472133874e9a477dc41def7df/src/pyhf/workspace.py#L303-L310
(validation happens at the bottom).

If there is no strong reason for the order being like this, I think it would be useful to run the validation first to filter out problematic specifications.

I ran into this when building a multi-channel workspace from a single-channel workspace for debugging purposes, and ended up pasting the channel at the wrong level of indentation.

### OS / Environment

```console
n/a
```

### Steps to Reproduce

```python (paste below)
import pyhf

spec = {
    "channels": [
        {
            "name": "SR",
            "samples_wrong_name": []
        }
    ]
}

pyhf.Workspace(spec)
```

### File Upload (optional)

_No response_

### Expected Results

`pyhf.exceptions.InvalidSpecification` or similar raised via the schema validation

### Actual Results

```console
Traceback (most recent call last):
  File "[...]/test.py", line 12, in <module>
    pyhf.Workspace(spec)
  File "[...]/pyhf/src/pyhf/workspace.py", line 303, in __init__
    super().__init__(spec, channels=spec['channels'])
  File "[...]/pyhf/src/pyhf/mixins.py", line 28, in __init__
    self.channel_nbins[channel['name']] = len(channel['samples'][0]['data'])
KeyError: 'samples'
```

### pyhf Version

```console
pyhf, version 0.7.0rc2.dev18
```

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by alexander-held**
# Description

The validation of a workspace specification against the schema currently takes place after some aspects of the workspace are already accessed in the constructor of `_ChannelSummaryMixin`. This can lead to unintuitive error messages, which should instead be caught by the schema validation. This PR re-orders the logic such that validation takes place first, ensuring that the relevant information is available for subsequent steps (assuming validation succeeds).

resolves #1952

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
* Validate the workspace specification in Model construction before any spec information
  can be accessed in the constructor of _ChannelSummaryMixin.
* Add test that checks for an pyhf.exceptions.InvalidSpecification being raised, indicating
  that the invalid model spec has been caught before a KeyError would be raised later.
```

### CODE REVIEW

File: tests/test_workspace.py [hunk 1]
```diff
@@ -910,3 +910,10 @@ def test_workspace_without_validation(mocker, simplemodels_model_data):
 
     pyhf.Workspace(dict(ws), validate=False)
     assert pyhf.schema.validate.called is False
+
+
+def test_workspace_invalid_specification():
+    spec = {"channels": [{"name": "SR", "samples_wrong_name": []}]}
+    # ensure that an invalid specifications gets caught as such
+    with pytest.raises(pyhf.exceptions.InvalidSpecification):
+        pyhf.Workspace(spec)
```
Comment (alexander-held) [inline tests/test_workspace.py:919]: This would fail before the changes in this PR, as it previously raised a `KeyError: 'samples'`.

**Review by kratsg — APPROVED** [review]

File: tests/test_workspace.py [hunk 2]
```diff
@@ -910,3 +910,10 @@ def test_workspace_without_validation(mocker, simplemodels_model_data):
 
     pyhf.Workspace(dict(ws), validate=False)
     assert pyhf.schema.validate.called is False
+
+
+def test_workspace_invalid_specification():
+    spec = {"channels": [{"name": "SR", "samples_wrong_name": []}]}
+    # ensure that an invalid specifications gets caught as such
```
Comment (matthewfeickert) [inline tests/test_workspace.py:917]:
```suggestion
    # Ensure that an invalid specifications gets caught as such
    # before a KeyError: 'samples' could be reached.
```

Comment (matthewfeickert) [reply tests/test_workspace.py:917]: This is all great, but I'd just add a comment to make it clear in the future what this test adds.

**Review by matthewfeickert — APPROVED** [review]

**Comment by matthewfeickert** [comment]
Thanks @alexander-held! We very much appreciate your contributions. :)

**Pull request merged**